modules/core/src/test/java/org/apache/ignite/internal/processors/compute/IgniteComputeWithPendingTopologyDeadlockTest.java

```diff
@@ -0,0 +1,304 @@
+            final IgniteCache<Integer, Integer> cache2 = node.createCache(cacheConfiguration(CACHE2, cacheMode, cacheWriteMode));
+
+            // Fill with test values.
+            fillCaches(ImmutableList.of(cache1, cache2));
+
+            final CountDownLatch sync = new CountDownLatch(1);
+
+            IgniteInternalFuture<Boolean> fut = runAsync(new Callable<Boolean>() {
+
+                @Override public Boolean call() throws IgniteInterruptedCheckedException {
+
+                    Lock lock = cache1.lock(2);
+
+                    lock.lock();
+
+                    try {
+                        // Signal main thread.
+                        sync.countDown();
+
+                        // Await when new cache will be created.
+                        boolean done = GridTestUtils.waitForCondition(new PA() {
+                            @Override public boolean apply() {
+                                return node.cacheNames().size() >= 3;
+                            }
+                        }, CACHE_CREATION_TIMEOUT);
+
+                        assertTrue(done);
+
+                        // Execute distributed (topology version aware) task.
+                        cache2.size();
+                    }
+                    catch (CacheException ignore) {
+                        // Expected behavior - computation was aborted.
+                    }
+                    finally {
+                        lock.unlock();
+                    }
+                    return true;
+                }
+            });
+
+            // Create watch-thread that will cancel task if timeout expired.
+            IgniteInternalFuture<Boolean> trackingFuture = startTracking(fut);
+
+            // Await new thread.
+            sync.await(PENDING_TIMEOUT, TimeUnit.MILLISECONDS);
+
+            // Create cache - topology update.
+            node.createCache(CACHE3);
+
+            // Await completion or timeout.
+            return trackingFuture.get();
+        }
+    }
+
+
+    /**
+     * Fill cache with test keys.
+     *
+     * @param caches caches to fill with test entries.
+     */
+    private void fillCaches(List<IgniteCache<Integer, Integer>> caches) {
+        for (IgniteCache<Integer, Integer> cache : caches)
+            for (int i = 0; i < CACHE_SIZE; i++)
+                cache.put(i, i);
+    }
+
+    /**
+     * Start timeout tracking.
+     *
+     * @param fut {@code IgniteInternalFuture} to track.
+     * @return {@code IgniteInternalFuture<Boolean>} that will return timeout status.
+     */
+    private IgniteInternalFuture<Boolean> startTracking(final IgniteInternalFuture<Boolean> fut) {
+        return runAsync(new Callable<Boolean>() {
+            @Override public Boolean call() {
+                try {
+                    // Busy wait for completion.
+                    boolean done = GridTestUtils.waitForCondition(new PA() {

```

Comment:
Why we can't use `fut.get(timeout)'` here?

Reply:
I didn't think about it :+1: